### PR TITLE
Added runexamples.sh script for automatic running of example python files

### PR DESCRIPTION
**Pull request by arp95**
@wyli 
This is the bash script i have worked till now. Is this way of running files fine? I have written till now for examples/classification_3d and examples/classification_3d_ignite files.
Also changed the model save paths so that I can run the train files first and then the eval files in those two folders.

part of #872 initial script for running the examples

### CODE REVIEW

**Comment by wyli** [comment]
thanks, looks like a very nice starting point for addressing https://github.com/Project-MONAI/MONAI/issues/872, could you please review @Nic-Ma

**Comment by arp95** [comment]
@wyli thanks. will be adding similar script code for other python files then.

**Comment by Nic-Ma** [comment]
Hi @wyli , I think the overall idea is good.
Should we add some logic to detect errors or compare results? And early stop if some example failed?

Thanks.

**Comment by wyli** [comment]
> Hi @wyli , I think the overall idea is good.
> Should we add some logic to detect errors or compare results? And early stop if some example failed?
> 
> Thanks.

yes, I think we could dump the demo logs to a temp file and start a python utility function to check the outputs strings? @arp95 
btw the device changes in this PR could be `"cuda" if torch.cuda.is_available() else "cpu"`

**Comment by arp95** [comment]
@wyli yes, updating that. so I will update the script for all the python files and then @Nic-Ma can  review and provide feedback.

**Comment by arp95** [comment]
@Nic-Ma can you please review the script till now. Then we can work on the error handling and other checks as you mentioned above.

File: examples/classification_3d/densenet_evaluation_array.py [hunk 1]
```diff
@@ -52,10 +52,10 @@ def main():
     val_loader = DataLoader(val_ds, batch_size=2, num_workers=4, pin_memory=torch.cuda.is_available())
 
     # Create DenseNet121
-    device = torch.device("cuda:0")
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
     model = monai.networks.nets.densenet.densenet121(spatial_dims=3, in_channels=1, out_channels=2).to(device)
 
-    model.load_state_dict(torch.load("best_metric_model.pth"))
+    model.load_state_dict(torch.load("best_metric_model_1.pth"))
```
Comment (Nic-Ma) [inline examples/classification_3d/densenet_evaluation_array.py:58]: Why change the name to "_1"? If you want to distinguish dict and array models, maybe use a more clear model name?

File: examples/classification_3d/densenet_evaluation_dict.py [hunk 2]
```diff
@@ -60,10 +60,10 @@ def main():
     val_loader = DataLoader(val_ds, batch_size=2, num_workers=4, pin_memory=torch.cuda.is_available())
 
     # Create DenseNet121
-    device = torch.device("cuda:0")
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
     model = monai.networks.nets.densenet.densenet121(spatial_dims=3, in_channels=1, out_channels=2).to(device)
 
-    model.load_state_dict(torch.load("best_metric_model.pth"))
+    model.load_state_dict(torch.load("best_metric_model_2.pth"))
```
Comment (Nic-Ma) [inline examples/classification_3d/densenet_evaluation_dict.py:66]: Why change the name to "_2"?

File: examples/classification_3d/densenet_training_array.py [hunk 3]
```diff
@@ -123,7 +123,7 @@ def main():
                 if metric > best_metric:
                     best_metric = metric
                     best_metric_epoch = epoch + 1
-                    torch.save(model.state_dict(), "best_metric_model.pth")
+                    torch.save(model.state_dict(), "best_metric_model_1.pth")
```
Comment (Nic-Ma) [inline examples/classification_3d/densenet_training_array.py:126]: same as the others.

File: examples/synthesis/gan_evaluation.py [hunk 4]
```diff
@@ -39,10 +40,10 @@ def main():
     monai.config.print_config()
     logging.basicConfig(stream=sys.stdout, level=logging.INFO)
     set_determinism(12345)
-    device = torch.device("cuda:0")
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
 
     # load generator
-    network_filepath = "./network_final.pth"
+    network_filepath = glob("./model_out/*.pth")[0]
```
Comment (Nic-Ma) [inline examples/synthesis/gan_evaluation.py:46]: Why change the model name?

File: examples/workflows/unet_evaluation_dict.py [hunk 5]
```diff
@@ -56,6 +56,9 @@ def main():
     segs = sorted(glob(os.path.join(tempdir, "seg*.nii.gz")))
     val_files = [{"image": img, "label": seg} for img, seg in zip(images, segs)]
 
+    # model file path
+    model_file = glob("./runs/net_key_metric*")[0]
```
Comment (Nic-Ma) [inline examples/workflows/unet_evaluation_dict.py:60]: If you want to select the best model for evaluation, should we `sort` the model names first?

File: examples/classification_3d/densenet_evaluation_array.py (hunk 1 above)
Comment (arp95) [reply examples/classification_3d/densenet_evaluation_array.py:58]: sure. i will update this along with other similar ones.

File: examples/synthesis/gan_evaluation.py (hunk 4 above)
Comment (arp95) [reply examples/synthesis/gan_evaluation.py:46]: the reason here was, when i ran the training file it saved the model in "model_out" folder. There was no './network_final.pth' model file available on running the training script. Can you please check that, if i did that right?

File: examples/workflows/unet_evaluation_dict.py (hunk 5 above)
Comment (arp95) [reply examples/workflows/unet_evaluation_dict.py:60]: no, only one file is being saved but each time with different metric. I mean earlier it was hardcoded with "./runs/net_key_metric=0.9101.pth" but in my case after running the training script, the model file saved was "./runs/net_key_metric=0.9305.pth". So as it might be different i am taking the regex form as done above.

File: examples/synthesis/gan_evaluation.py (hunk 4 above)
Comment (Nic-Ma) [reply examples/synthesis/gan_evaluation.py:46]: Hi @gagandaroach ,

Could you please help double confirm the saved model name in the GAN example? Seems the `training` and `evaluation` programs mismatch?

Thanks.

File: examples/workflows/unet_evaluation_dict.py (hunk 5 above)
Comment (Nic-Ma) [reply examples/workflows/unet_evaluation_dict.py:60]: I see, then it looks good to me.
thanks.

**Comment by Nic-Ma** [comment]
Hi @arp95 ,

Thanks for your quick update.
I think you can go for next steps with error check and early stop now.

Thanks.

**Comment by arp95** [comment]
Thanks for the review @Nic-Ma . Will start working on the error check and early stop.

File: examples/synthesis/gan_evaluation.py (hunk 4 above)
Comment (gagandaroach) [reply examples/synthesis/gan_evaluation.py:46]: Thanks @arp95 for fixing this. 

For the `evaluation` script I would manually copy the final checkpoint from `trainer` into the base directory as `./network_final.pth`. The proposed changes automatically grab a `.pth` to execute  `evaluation`. 

May I suggest the test ensure `training` went to completion by checking if the final checkpoint is saved. `glob("./model_out/checkpoint_final*.pth")` here in `evaluation`.

Comment (arp95) [reply examples/synthesis/gan_evaluation.py:46]: thanks @gagandaroach Yes will add that test in the script.

**Comment by arp95** [comment]
@Nic-Ma i have added code for early stopping of the bash script and error checking.

1. "set -e" and "exit 0" ensure we exit the bash script if the python file doesnt execute properly or the model file isnt generated from the training script due to some reason.
2. If python file doesnt execute properly, "set -e" ensures no further processing takes place and we exit from the script.
3. Only if the required model file is generated from the training file, I go on to the eval files in that folder and execute them.
4. Also, the "temp.txt" is created where I am outputting "1" or "0" which can be used for further analysis to check success or failure.

Is this going right?

**Comment by Nic-Ma** [comment]
Hi @arp95 ,

Thanks for your quick update!
I have 4 minor comments:
1. Before downloading the IXI dataset, please check whether it's already existing, if yes, skip downloading.
2. Same as extracting, if already extracted the zip file, skip extracting.
3. I think maybe you can add more information to `temp` file instead of only "0", and it's better to print the same error message on screen.
4. Please add the steps to install necessary packages to run examples, like `pip install monai[nibabel]`, etc.

Thanks.

**Comment by arp95** [comment]
thanks @Nic-Ma . Working on adding these points.

**Comment by arp95** [comment]
@Nic-Ma please check now.

**Comment by arp95** [comment]
@wyli can you please check and update on what should be done ahead? Thanks.

**Comment by wyli** [comment]
> @wyli can you please check and update on what should be done ahead? Thanks.

Hi @arp95 I tried the script,  the package installation and data downloading work fine. But the script stops too early and doesn't run all the commands, could you help check? also I think we should clean up `temp.txt` when starting the script

**Comment by arp95** [comment]
@wyli sure looking into it.

**Comment by arp95** [comment]
@wyli please check now.

**Review by wyli — APPROVED** [review]
thanks, the script can go through all the examples, still need to improve the checking of the examples' outputs

**Comment by arp95** [comment]
@wyli sure. will look into it.
@Nic-Ma any ideas on how i can improve the checking of examples' outputs? Thanks.
